rosbag2_py/src/rosbag2_py/_transport.cpp

```diff
@@ -156,20 +171,28 @@ class Recorder
     }
 
 
-    auto reader = std::make_shared<rosbag2_cpp::Reader>(
-      std::make_unique<rosbag2_cpp::readers::SequentialReader>());
-    std::shared_ptr<rosbag2_cpp::Writer> writer;
+    std::unique_ptr<rosbag2_cpp::Writer> writer = nullptr;
     // Change writer based on recording options
     if (!record_options.compression_format.empty()) {
-      writer = std::make_shared<rosbag2_cpp::Writer>(
+      writer = std::make_unique<rosbag2_cpp::Writer>(
         std::make_unique<rosbag2_compression::SequentialCompressionWriter>(compression_options));
     } else {
-      writer = std::make_shared<rosbag2_cpp::Writer>(
+      writer = std::make_unique<rosbag2_cpp::Writer>(
         std::make_unique<rosbag2_cpp::writers::SequentialWriter>());
     }
 
-    rosbag2_transport::Recorder impl(writer);
-    impl.record(storage_options, record_options);
+    auto recorder = std::make_shared<rosbag2_transport::Recorder>(
+      std::move(writer), storage_options, record_options);
+    recorder->record();
+    rclcpp::executors::SingleThreadedExecutor exec;
+    exec.add_node(recorder);
+    auto spin_thread = std::thread(
```

Comment:
Don't you want to start spinning before calling record? IIRC `record` doesn't exit until it's cancelled.

Reply:
hmm. I honestly don't know. I had the same question at #741 within the tests - this code is actually really just copied from these. 
I thought it would make sense to give `record` the change to setup all subscriptions and potential error handling and really only spin once that's completed. 

Why do you advocate for spinning first?

Reply:
Oh - `Recorder::record` returns right away after setup and runs in the background, until the `Recorder` object is destroyed. API confusion from the previous blocking API. 

Not something we have to handle here, but maybe we ought to consider a slightly more obvious API - if recording is controlled by the object lifecycle, then should recording just start in the constructor? Or, alternatively should there be a way to stop recording without destroying the object?